src/rebar_pkg_resource.erl

```diff
@@ -206,17 +206,19 @@ store_etag_in_cache(Path, ETag) ->
 %%%=============================================================================
 %%% Private functions
 %%%=============================================================================
--spec cached_download(TmpDir, CachePath, Pkg, ETag, ETagPath, UpdateETag) -> Res when
+-spec cached_download(TmpDir, CachePath, Pkg, State, ETag, ETagPath, UpdateETag) -> Res when
       TmpDir :: file:name(),
       CachePath :: file:name(),
       Pkg :: package(),
+      State :: rebar_state:t(),
       ETag :: binary(),
       ETagPath :: file:name(),
       UpdateETag :: boolean(),
       Res :: ok | {unexpected_hash, integer(), integer()} | {fetch_fail, binary(), binary()}.
-cached_download(TmpDir, CachePath, Pkg={pkg, Name, Vsn, _Hash, RepoConfig}, ETag,
+cached_download(TmpDir, CachePath, Pkg={pkg, Name, Vsn, _Hash, RepoConfig}, State, ETag,
                 ETagPath, UpdateETag) ->
-    case request(RepoConfig, Name, Vsn, ETag) of
+    CDN = rebar_state:get(State, rebar_packages_cdn, ?DEFAULT_CDN),
```

Comment:
@ferd Had to change this back to get/3 per tests breaking.

Reply:
yeah I guess the context plays here. It might actually be safest to keep the old format (lists for URLs) and convert closest to the call-site (the actual request) so we don't explode plugins that might hook into these semi-private interfaces either. Sorry for the back and forth there. I guess it just _is_ simpler to convert to the needed implementation-specific format at the call-site.

Reply:
No worries, back and forth is how it goes sometimes, as is tradition. And yes, I agree.

Reply:
Let me throw out an idea here though. What we could do to keep these details  isolated to a more proper place we could add a function on `rebar_hex_repos` such as `set_cdn/2` which could take a Repo and State. Perhaps better would be to call it `config_for_download/2`.

Reply:
Maybe. I would tend to avoid relying on State for deeper specialized interfaces though, since they make testing a bit trickier and make some specific code attached to the rebar3 context without needing it.

Reply:
Are you suggesting not putting this in state at all and instead reaching out to env for a possibly hex cdn value at the callsite?

Reply:
No I'm saying I'm fine with the current form. Just probably go back to storing as a list and convert to binary when sending it to the request call.

Reply:
aye, done.